### PR TITLE
Handle metrics during schema transformation

### PR DESCRIPTION
**Pull request by holdenk**
This is for #2079, when we transform the schema we should update the metrics as well automatically.
Longer term we may wish to consider also migrating the metrics to column references, but either way for backwards compatibility we should preserve the name based fields.

### CODE REVIEW

**Comment by holdenk** [comment]
cc @aokolnychyi

**Comment by aokolnychyi** [comment]
Let me take a look today! Thanks, @holdenk!

**Comment by aokolnychyi** [comment]
Sorry, I won't get to this one today. I'll review it tomorrow.

cc @RussellSpitzer

File: core/src/main/java/org/apache/iceberg/SchemaUpdate.java [hunk 1]
```diff
@@ -382,27 +378,43 @@ private int assignNewColumnId() {
...
+        LOG.warn("Failed to update external schema mapping: {}", existingMappingJson, e);
       }
     }
 
-    return metadata;
+    // Transform the metrics if they exist
+    if (base != null && base.properties() != null) {
+      Schema newSchema = metadata.schema();
+      List<String> deletedColumns = deletes.stream()
+          .map(i -> schema.findColumnName(i))
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/SchemaUpdate.java:406]: a nit but I would rather it was something like "name" instead of i but it may be my bias that I assume i is an int.

File: core/src/test/java/org/apache/iceberg/TestSchemaAndMappingUpdate.java [hunk 2]
```diff
@@ -176,20 +176,11 @@ public void testModificationWithMetricsMetrics() {
...
-    AssertHelpers.assertThrows(
-        "Deleting a column with metrics fails",
-        ValidationException.class,
-        null,
-        () ->
-          table.updateSchema()
-              .deleteColumn("id")
-              .commit());
+    // Re-naming a column with metrics succeeds;
+    table.updateSchema().renameColumn("id", "bloop").commit();
```
Comment (RussellSpitzer) [inline core/src/test/java/org/apache/iceberg/TestSchemaAndMappingUpdate.java:180]: I think we may also want to validate that the metric config is copied to the new name correctly, and deleted in the example below. Just to make sure the same value goes through

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java [hunk 3]
```diff
@@ -46,6 +47,35 @@ public static MetricsConfig getDefault() {
     return spec;
   }
 
+  public static Map<String, String> updateProperties(Map<String, String> props, List<String> deletes,
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetricsConfig.java:50]: Seems like we don't need it as public?

I'd also consider a utility class `MetricsConfigUtil` to keep `MetricsConfig` simple. We will need a few helper methods for metrics configs and it may be a good time to introduce a utility class now. That way, we won't have to reason about compatibility when we move to column ids.

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java [hunk 4]
```diff
@@ -46,6 +47,35 @@ public static MetricsConfig getDefault() {
...
+                                                     Map<String, String> renames) {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetricsConfig.java:51]: nit: I think `deletedColumns` and `renamedColumns` would be a bit better. That's exactly the variable names we pass in from `SchemaUpdate`.

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java [hunk 5]
```diff
@@ -46,6 +47,35 @@ public static MetricsConfig getDefault() {
...
+    if (!props.keySet().stream().anyMatch(key -> key.startsWith(TableProperties.METRICS_MODE_COLUMN_CONF_PREFIX))) {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetricsConfig.java:52]: nit: shall we do a static import for `METRICS_MODE_COLUMN_CONF_PREFIX`? We have two such imports in this class already.

Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetricsConfig.java:52]: I think using `noneMatch` instead of `anyMatch` would allow us to remove the negation, making this a bit easier.

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java [hunk 6]
```diff
@@ -46,6 +47,35 @@ public static MetricsConfig getDefault() {
...
+      return props;
+    } else {
+      Map<String, String> updatedProperties = Maps.newHashMap();
+      // Put all of the non metrics columns we aren't modifying
+      props.keySet().stream().forEach(key -> {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetricsConfig.java:57]: nit: could be simply `props.keySet().forEach`

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java [hunk 7]
```diff
@@ -46,6 +47,35 @@ public static MetricsConfig getDefault() {
...
+        if (key.startsWith(TableProperties.METRICS_MODE_COLUMN_CONF_PREFIX)) {
+          String columnAlias = key.replaceFirst(TableProperties.METRICS_MODE_COLUMN_CONF_PREFIX, "");
+          if (deletes.contains(columnAlias)) {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetricsConfig.java:60]: I think we can reduce the number of branches here:

```
if (renames.containsKey(columnAlias)) {
  // the name has changed
  ...
 } else if (!deletes.contains(columnAlias)) {
   // copy over the config as it has not changed
  ...
}
```

File: core/src/main/java/org/apache/iceberg/SchemaUpdate.java [hunk 8]
```diff
@@ -382,27 +378,43 @@ private int assignNewColumnId() {
   }
 
   private TableMetadata applyChangesToMapping(TableMetadata metadata) {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/SchemaUpdate.java:380]: This method has a very specific name as it was previously only handling updates to the name mapping. I think we should rename it into something more generic now. What about just `applyChanges`?

File: core/src/main/java/org/apache/iceberg/SchemaUpdate.java [hunk 9]
```diff
@@ -382,27 +378,43 @@ private int assignNewColumnId() {
...
-    String mappingJson = metadata.property(TableProperties.DEFAULT_NAME_MAPPING, null);
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/SchemaUpdate.java:385]: Do we need to rename `mappingJson`? Seems like it is optional so may be a good idea to avoid unnecessary changes to simplify cherry-picks and reviews.

File: core/src/main/java/org/apache/iceberg/SchemaUpdate.java [hunk 10]
```diff
@@ -382,27 +378,43 @@ private int assignNewColumnId() {
...
-    if (mappingJson != null) {
+    String existingMappingJson = metadata.property(TableProperties.DEFAULT_NAME_MAPPING, null);
+    TableMetadata ret = metadata;
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/SchemaUpdate.java:382]: Will `newMetadata` be a better option?

File: core/src/main/java/org/apache/iceberg/SchemaUpdate.java (hunk 1 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/SchemaUpdate.java:406]: +1

We can even simplify it further, I guess:

```
deletes.stream()
   .map(schema::findColumnName)
   .collect(Collectors.toList());
```

**Comment by aokolnychyi** [comment]
The approach in this PR looks correct, I only had cosmetic comments.

File: core/src/main/java/org/apache/iceberg/SchemaUpdate.java (hunk 1 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/SchemaUpdate.java:406]: I think `i` means a column id when we build `renamedColumns`. Maybe, just using `id` would make sense.

```
updates.keySet().stream()
     .filter(id -> !schema.findColumnName(id).equals(newSchema.findColumnName(id)))
     .collect(Collectors.toMap(schema::findColumnName, newSchema::findColumnName));
```

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java (hunk 3 above)
Comment (holdenk) [reply core/src/main/java/org/apache/iceberg/MetricsConfig.java:50]: I think this needs to be public since it's called from another class no?

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java [hunk 11]
```diff
@@ -87,4 +117,8 @@ public void validateReferencedColumns(Schema schema) {
   public MetricsMode columnMode(String columnAlias) {
     return columnModes.getOrDefault(columnAlias, defaultMode);
   }
+
+  public boolean allDefault() {
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/MetricsConfig.java:121]: Not sure where this is used?

Comment (holdenk) [reply core/src/main/java/org/apache/iceberg/MetricsConfig.java:121]: Good catch, I don't remember why I added this myself it's been so long. But yeah I don't see any callers so I'll take it out.

File: core/src/test/java/org/apache/iceberg/TestSchemaAndMappingUpdate.java (hunk 2 above)
Comment (holdenk) [reply core/src/test/java/org/apache/iceberg/TestSchemaAndMappingUpdate.java:180]: Added :)

**Review by RussellSpitzer — APPROVED** [review]

**Comment by RussellSpitzer** [comment]
Looks good to me, @aokolnychyi Were you 👍 ?

**Comment by aokolnychyi** [comment]
Sorry for the delay, let me take a look once I get out of meetings.

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java (hunk 3 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MetricsConfig.java]: This class as well as `SchemaUpdate` (where the new method is used) are in `org.apache.iceberg` package. I don't think users are supposed to interact with this logic, so I'd make the method package-private like we do in other places.

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java [hunk 12]
```diff
@@ -46,6 +47,35 @@ public static MetricsConfig getDefault() {
...
+      return props;
+    } else {
+      Map<String, String> updatedProperties = Maps.newHashMap();
+      // Put all of the non metrics columns we aren't modifying
+      props.keySet().forEach(key -> {
+        if (key.startsWith(TableProperties.METRICS_MODE_COLUMN_CONF_PREFIX)) {
+          String columnAlias = key.replaceFirst(TableProperties.METRICS_MODE_COLUMN_CONF_PREFIX, "");
+          if (renamedColumns.get(columnAlias) != null) {
+            // The name has changed.
+            String newKey = TableProperties.METRICS_MODE_COLUMN_CONF_PREFIX +
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetricsConfig.java:62]: Can we add a static import for `METRICS_MODE_COLUMN_CONF_PREFIX` like we have for other two props in this class? That way, the class will be a little bit more consistent and this logic will fit on one line.

File: core/src/main/java/org/apache/iceberg/SchemaUpdate.java (hunk 9 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/SchemaUpdate.java:405]: I'd still skip the rename if possible.

File: core/src/main/java/org/apache/iceberg/SchemaUpdate.java [hunk 13]
```diff
@@ -401,28 +396,43 @@ private int assignNewColumnId() {
...
         // log the error, but do not fail the update
-        LOG.warn("Failed to update external schema mapping: {}", mappingJson, e);
+        LOG.warn("Failed to update external schema mapping: {}", existingMappingJson, e);
       }
     }
 
-    return metadata;
+    // Transform the metrics if they exist
+    if (base != null && base.properties() != null) {
+      Schema newSchema = metadata.schema();
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/SchemaUpdate.java:423]: nit: it probably does not matter now but shall we use `newMetadata` here in case if the logic above changes?

**Comment by aokolnychyi** [comment]
Just a few comments/questions. Otherwise, looks good to go. Thanks, @holdenk!

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java [hunk 14]
```diff
@@ -46,6 +48,34 @@ public static MetricsConfig getDefault() {
     return spec;
   }
 
+  static Map<String, String> updateProperties(Map<String, String> props, List<String> deletedColumns,
+                                                     Map<String, String> renamedColumns) {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetricsConfig.java:52]: nit: arg formatting

**Review by aokolnychyi — APPROVED** [review]
LGTM

**Comment by aokolnychyi** [comment]
@holdenk, could you please fix the formatting [here](https://github.com/apache/iceberg/pull/2257/files#r634636704)? Let's get this in.

**Comment by aokolnychyi** [comment]
Thanks, @holdenk! Thanks for reviewing, @RussellSpitzer!
